### PR TITLE
Sassafras primitives

### PR DESCRIPTION
**Pull request by davxy**
This PR introduces a new crate with the primitives required by https://github.com/paritytech/polkadot-sdk/issues/41.

Quoting the crate [README](https://github.com/davxy/substrate/blob/sassafras-primitives/primitives/consensus/sassafras/README.md):

    The crate interfaces and structures are experimental and may be subject to changes.

Nevertheless, the strategy here is to incrementally drain the main Sassafras PR by upstreaming smaller components in order to make the overall protocol inclusion and analysis more manageable.

The structures and functions were found to satisfy the requirements of the protocol.

A more in depth analysis of the structures obviously requires to dig into the client and frame pallets in the main PR.

---

This crate requires to enable bandersnatch-experimental feature in some of its dependencies.

This feature ends up being transitively enabled in:
sp-consensus-sassafras → sp-application-crypto → sp-io → sp-keystore

Follows that Bandersnatch API ends up being exposed by the `Keystore` trait and thus we require to expose these functions implementations in the client keystore (`sc-keystore`) as well.

The (temporary) ~workaround is to include `sc-keystore` as an optional dependency in `sp-consensus-sassafras` and enable it with `bandersnatch-experimental` feature on std builds~ solution is to conditionally compile `sc-keystore` experimental code using the result of `generate_feature_enabled_macro`  (see https://github.com/paritytech/polkadot-sdk/pull/1249/commits/3a1eedae9bcd56fd9f5b8ed181d1672e9d486ac1)

---

Step towards: https://github.com/paritytech/polkadot-sdk/issues/41

Extracted from: https://github.com/paritytech/substrate/pull/11879

Superseeds: https://github.com/paritytech/substrate/pull/14768

### CODE REVIEW

**Review by skunert — APPROVED** [review]
Based on your previous explanations about the flow, this looks good to me!

File: substrate/primitives/consensus/sassafras/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+# Sassafras requires to enable `bandersnatch-experimental` feature for some of its dependencies
+# and this ends up being transitively enabled in `sp-keystore` crate.
+#
+# E.g. sp-application-crypto → sp-io → sp-keystore
+#
+# Follows that Bandersnatch api ends up being exposed by the `Keystore` trait and thus we require
+# to expose these functions implementations in the client keystore (`sc-keystore`) as well.
+#
+# The crate feature should be enabled only for `std` builds (keep it disabled by default).
+sc-keystore = { path = "../../../client/keystore", optional = true}
```
Comment (bkchr) [inline substrate/primitives/consensus/sassafras/Cargo.toml:39]: https://github.com/paritytech/polkadot-sdk/blob/7d77005522e4138eef9a7b55850ef4a5aa72d8ff/substrate/primitives/core/src/lib.rs#L404-L424 use this macro to create a `bandersnatch_enabled` macro to guard the implementation of the trait methods in `sc-keystore`.

**Review by bkchr — APPROVED** [review]

File: substrate/primitives/consensus/sassafras/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+/// An equivocation proof for multiple block authorships on the same slot (i.e. double vote).
+pub type EquivocationProof<H> = sp_consensus_slots::EquivocationProof<H, AuthorityId>;
+
+/// Randomness required by some protocol's operations.
+pub type Randomness = [u8; RANDOMNESS_LENGTH];
+
+/// Configuration data that can be modified on epoch change.
+#[derive(
+	Copy, Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug, MaxEncodedLen, TypeInfo, Default,
```
Comment (koute) [inline substrate/primitives/consensus/sassafras/src/lib.rs:85]: Creating this struct with `Default` will initialize everything to zero, including the `attempts_number`. Shouldn't that be at least `1` by default?

File: substrate/primitives/consensus/sassafras/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+/// An opaque type used to represent the key ownership proof at the runtime API boundary.
+///
+/// The inner value is an encoded representation of the actual key ownership proof which will be
+/// parameterized when defining the runtime. At the runtime API boundary this type is unknown and
+/// as such we keep this opaque representation, implementors of the runtime API will have to make
+/// sure that all usages of `OpaqueKeyOwnershipProof` refer to the same type.
+#[derive(Decode, Encode, PartialEq, TypeInfo)]
+pub struct OpaqueKeyOwnershipProof(Vec<u8>);
```
Comment (koute) [inline substrate/primitives/consensus/sassafras/src/lib.rs:121]:
```suggestion
#[derive(Decode, Encode, PartialEq, TypeInfo)]
#[repr(transparent)]
pub struct OpaqueKeyOwnershipProof(Vec<u8>);
```

**Review by koute — APPROVED** [review]

File: substrate/primitives/consensus/sassafras/Cargo.toml (hunk 1 above)
Comment (davxy) [reply substrate/primitives/consensus/sassafras/Cargo.toml]: I wasn't aware of this. Is doing the job better. Thank you

**Review by michalkucharczyk — APPROVED** [review]

File: substrate/primitives/consensus/sassafras/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+			slot: Slot,
+			authority_id: AuthorityId,
+		) -> Option<OpaqueKeyOwnershipProof>;
+
+		/// Submits an unsigned extrinsic to report an equivocation.
+		///
+		/// The caller must provide the equivocation proof and a key ownership proof (should be
+		/// obtained using `generate_key_ownership_proof`). The extrinsic will be unsigned and
+		/// should only be accepted for local authorship (not to be broadcast to the network). This
+		/// method returns `None` when creation of the extrinsic fails, e.g. if equivocation
```
[the remark this answers is not in the document]
Comment (davxy) [reply substrate/primitives/consensus/sassafras/src/lib.rs]: https://github.com/paritytech/polkadot-sdk/pull/1249/commits/c17ae1bf20a88a63b4e6744bd13ce0ecd1039a02

File: substrate/primitives/consensus/sassafras/src/lib.rs (hunk 3 above)
Comment (davxy) [reply substrate/primitives/consensus/sassafras/src/lib.rs]: https://github.com/paritytech/polkadot-sdk/pull/1249/commits/c17ae1bf20a88a63b4e6744bd13ce0ecd1039a02

**Pull request merged**